File: src/tool/omniidl/python/omniidl/idltype.py
#
# IDL type representation
#

import idlutil

tk_null               = 0
tk_void               = 1
tk_short              = 2
tk_long               = 3
tk_ushort             = 4
tk_ulong              = 5
tk_float              = 6
tk_double             = 7
tk_boolean            = 8
tk_char	              = 9
tk_octet              = 10
tk_any	              = 11
tk_TypeCode           = 12
tk_Principal          = 13
tk_objref             = 14
tk_struct             = 15
tk_union              = 16
tk_enum	              = 17
tk_string             = 18
tk_sequence           = 19
tk_array              = 20
tk_alias              = 21
tk_except             = 22
tk_longlong           = 23
tk_ulonglong          = 24
tk_longdouble         = 25
tk_wchar              = 26
tk_wstring            = 27
tk_fixed              = 28
tk_value              = 29
tk_value_box          = 30
tk_native             = 31
tk_abstract_interface = 32


class Error:
    def __init__(self, err):
        self.err = err

    def __repr__(self):
        return self.err


class Type:
    def __init__(self, kind):
        self.__kind  = kind

    # Typecode kind
    def kind(self):  return self.__kind


# Base types
class Base (Type):
    def __init__(self, kind):
        if kind != tk_null       and \
           kind != tk_void       and \
           kind != tk_short      and \
           kind != tk_long       and \
           kind != tk_ushort     and \
           kind != tk_ulong      and \
           kind != tk_float      and \
           kind != tk_double     and \
           kind != tk_boolean    and \
           kind != tk_char       and \
           kind != tk_octet      and \
           kind != tk_any        and \
           kind != tk_TypeCode   and \
           kind != tk_Principal  and \
           kind != tk_longlong   and \
           kind != tk_ulonglong  and \
           kind != tk_longdouble and \
           kind != tk_wchar:
            raise Error("Attempt to create Base type with invalid kind.")

        Type.__init__(self, kind)

    def accept(self, visitor): visitor.visitBaseType(self)


# Strings can be used like base types without a declaration. eg:
#
#   void op(in string<10> s);
#
# therefore, the String type must include its bound here, rather than
# relying on looking at the corresponding declaration

class String (Type):
    def __init__(self, bound):
        Type.__init__(self, tk_string)
        self.__bound = bound

    def accept(self, visitor): visitor.visitStringType(self)
    def bound(self): return self.__bound

class WString (Type):
    def __init__(self, bound):
        Type.__init__(self, tk_string)
        self.__bound = bound

    def accept(self, visitor): visitor.visitWStringType(self)
    def bound(self): return self.__bound


# Sequences are never declared. They either appear as
#
#   typedef sequence <...> ...
#
# or inside a struct or union

class Sequence (Type):
    def __init__(self, seqType, bound):
        Type.__init__(self, tk_sequence)
        self.__seqType = seqType
        self.__bound   = bound

    def accept(self, visitor): visitor.visitSequenceType(self)
    def seqType(self): return self.__seqType
    def bound(self):   return self.__bound


# Same goes for fixed

class Fixed (Type):
    def __init__(self, digits, scale):
        Type.__init__(self, tk_fixed)
        self.__digits = digits
        self.__scale  = scale

    def accept(self, visitor): visitor.visitFixedType(self)
    def digits(self): return self.__digits
    def scale(self):  return self.__scale


# All other types must be declared, at least implicitly, so they have
# an associated declaration object

class Declared (Type):
    def __init__(self, decl, scopedName, kind):
        if kind != tk_objref             and \
           kind != tk_struct             and \
           kind != tk_union              and \
           kind != tk_enum               and \
           kind != tk_array              and \
           kind != tk_alias              and \
           kind != tk_except             and \
           kind != tk_value              and \
           kind != tk_value_box          and \
           kind != tk_abstract_interface:
            raise Error("Attempt to create Declared type with invalid kind.")

        Type.__init__(self, kind)
        self.__decl       = decl
        self.__scopedName = scopedName

    def accept(self, visitor): visitor.visitDeclaredType(self)

    # Decl object where the type was declared.
    def decl(self):       return self.__decl

    # List containing scoped name:
    def scopedName(self): return self.__scopedName

    # Simple name
    def name(self):       return self.__scopedName[-1:]


# Map of singleton Base Type objects
baseTypeMap = {
    tk_null:       Base(tk_null),
    tk_void:       Base(tk_void),
    tk_short:      Base(tk_short),
    tk_long:       Base(tk_long),
    tk_ushort:     Base(tk_ushort),
    tk_ulong:      Base(tk_ulong),
    tk_float:      Base(tk_float),
    tk_double:     Base(tk_double),
    tk_boolean:    Base(tk_boolean),
    tk_char:       Base(tk_char),
    tk_octet:      Base(tk_octet),
    tk_any:        Base(tk_any),
    tk_TypeCode:   Base(tk_TypeCode),
    tk_Principal:  Base(tk_Principal),
    tk_longlong:   Base(tk_longlong),
    tk_ulonglong:  Base(tk_ulonglong),
    tk_longdouble: Base(tk_longdouble),
    tk_wchar:      Base(tk_wchar)
    }

# Maps of String and WString Type objects, indexed by bound
stringTypeMap =  { 0: String(0) }
wstringTypeMap = { 0: WString(0) }

# Map of Sequence Type objects, indexed by (type object,bound)
sequenceTypeMap = {}

# Map of Fixed Type objects, indexed by (digits,scale)
fixedTypeMap = {}

# Map of declared type objects, indexed by stringified scoped name
declaredTypeMap = {}


# Functions to create or return existing Type objects
def baseType(kind):
    return baseTypeMap[kind]

def stringType(bound):
    try:
        return stringTypeMap[bound]
    except KeyError:
        st = String(bound)
        stringTypeMap[bound] = st
        return st

def wstringType(bound):
    try:
        return wstringTypeMap[bound]
    except KeyError:
        wst = WString(bound)
        wstringTypeMap[bound] = wst
        return wst

def sequenceType(type_spec, bound):
    try:
        return sequenceTypeMap[(type_spec,bound)]
    except KeyError:
        st = Sequence(type_spec, bound)
        sequenceTypeMap[(type_spec,bound)] = st
        return st

def fixedType(digits, scale):
    try:
        return fixedTypeMap[(digits,scale)]
    except KeyError:
        ft = Fixed(digits, scale)
        fixedTypeMap[(digits,scale)] = ft
        return ft

def declaredType(decl, scopedName, kind):
    sname = idlutil.slashName(scopedName)
    try:
        return declaredTypeMap[sname]
    except KeyError:
        dt = Declared(decl, scopedName, kind)
        declaredTypeMap[sname] = dt
        return dt

File: src/tool/omniidl/cxx/idltype.cc
// -*- c++ -*-
//                          Package   : omniidl
// idltype.cc               Created on: 1999/10/21
//			    Author    : Duncan Grisby (dpg1)
//
//    Copyright (C) 1999 AT&T Laboratories Cambridge
//
//  This file is part of omniidl.
//
//  omniidl is free software; you can redistribute it and/or modify it
//  under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//  General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
//  02111-1307, USA.
//
// Description:
//   
//   Type objects

// $Id$
// $Log$
// Revision 1.8  2000/07/13 15:25:52  dpg1
// Merge from omni3_develop for 3.0 release.
//
// Revision 1.5.2.1  2000/03/06 10:40:29  dpg1
// Typo in string constant.
//
// Revision 1.5  1999/11/17 17:16:59  dpg1
// Changes to remove static initialisation of objects.
//
// Revision 1.4  1999/11/04 17:16:54  dpg1
// Changes for NT.
//
// Revision 1.3  1999/11/02 17:07:24  dpg1
// Changes to compile on Solaris.
//
// Revision 1.2  1999/10/29 15:44:14  dpg1
// DeclaredType() now takes extra DeclRepoId* argument.
//
// Revision 1.1  1999/10/27 14:05:54  dpg1
// *** empty log message ***
//

#include <idltype.h>
#include <idlerr.h>

const char*
IdlType::
kindAsString() const
{
  switch(kind_) {
  case tk_null:               return "null";
  case tk_void:               return "void";
  case tk_short:              return "short";
  case tk_long:               return "long";
  case tk_ushort:             return "unsigned short";
  case tk_ulong:              return "unsigned long";
  case tk_float:              return "float";
  case tk_double:             return "double";
  case tk_boolean:            return "boolean";
  case tk_char:               return "char";
  case tk_octet:              return "octet";
  case tk_any:                return "any";
  case tk_TypeCode:           return "CORBA::TypeCode";
  case tk_Principal:          return "CORBA::Principal";
  case tk_objref:             return "interface";
  case tk_struct:             return "struct";
  case tk_union:              return "union";
  case tk_enum:               return "enum";
  case tk_string:             return "string";
  case tk_sequence:           return "sequence";
  case tk_array:              return "array";
  case tk_alias:              return "typedef";
  case tk_except:             return "exception";
  case tk_longlong:           return "long long";
  case tk_ulonglong:          return "unsigned long long";
  case tk_longdouble:         return "long double";
  case tk_wchar:              return "wchar";
  case tk_wstring:            return "wstring";
  case tk_fixed:              return "fixed";
  case tk_value:              return "value";
  case tk_value_box:          return "value box";
  case tk_native:             return "native";
  case tk_abstract_interface: return "abstract interface";
  }
  assert(0);
  return ""; // To keep MSVC happy
}

IdlType*
IdlType::
scopedNameToType(const char* file, int line, const ScopedName* sn)
{
  const Scope::Entry* se = Scope::current()->findForUse(sn, file, line);

  if (se) {
    if (se->kind() == Scope::Entry::E_DECL) {
      IdlType *t = se->idltype();
      if (t) return t;
    }
    char* ssn = sn->toString();
    IdlError(file, line, "`%s' is not a type", ssn);
    IdlErrorCont(se->file(), se->line(), "(`%s' declared here)", ssn);
    delete [] ssn;
  }
  return 0;
}

// Static type object pointers
_CORBA_Boolean IdlType::initialised_             = 0;
BaseType*      BaseType::nullType                = 0;
BaseType*      BaseType::voidType                = 0;
BaseType*      BaseType::shortType               = 0;
BaseType*      BaseType::longType                = 0;
BaseType*      BaseType::ushortType              = 0;
BaseType*      BaseType::ulongType               = 0;
BaseType*      BaseType::floatType               = 0;
BaseType*      BaseType::doubleType              = 0;
BaseType*      BaseType::booleanType             = 0;
BaseType*      BaseType::charType                = 0;
BaseType*      BaseType::octetType               = 0;
BaseType*      BaseType::anyType                 = 0;
BaseType*      BaseType::TypeCodeType            = 0;
BaseType*      BaseType::PrincipalType           = 0;
BaseType*      BaseType::longlongType            = 0;
BaseType*      BaseType::ulonglongType           = 0;
BaseType*      BaseType::longdoubleType          = 0;
BaseType*      BaseType::wcharType               = 0;
StringType*    StringType::unboundedStringType   = 0;
WStringType*   WStringType::unboundedWStringType = 0;
DeclaredType*  DeclaredType::corbaObjectType     = 0;


void
IdlType::
init()
{
  if (!initialised_) {
    BaseType::nullType                = new BaseType(IdlType::tk_null);
    BaseType::voidType                = new BaseType(IdlType::tk_void);
    BaseType::shortType               = new BaseType(IdlType::tk_short);
    BaseType::longType                = new BaseType(IdlType::tk_long);
    BaseType::ushortType              = new BaseType(IdlType::tk_ushort);
    BaseType::ulongType               = new BaseType(IdlType::tk_ulong);
    BaseType::floatType               = new BaseType(IdlType::tk_float);
    BaseType::doubleType              = new BaseType(IdlType::tk_double);
    BaseType::booleanType             = new BaseType(IdlType::tk_boolean);
    BaseType::charType                = new BaseType(IdlType::tk_char);
    BaseType::octetType               = new BaseType(IdlType::tk_octet);
    BaseType::anyType                 = new BaseType(IdlType::tk_any);
    BaseType::TypeCodeType            = new BaseType(IdlType::tk_TypeCode);
    BaseType::PrincipalType           = new BaseType(IdlType::tk_Principal);
    BaseType::longlongType            = new BaseType(IdlType::tk_longlong);
    BaseType::ulonglongType           = new BaseType(IdlType::tk_ulonglong);
    BaseType::longdoubleType          = new BaseType(IdlType::tk_longdouble);
    BaseType::wcharType               = new BaseType(IdlType::tk_wchar);
    StringType::unboundedStringType   = new StringType(0);
    WStringType::unboundedWStringType = new WStringType(0);
    DeclaredType::corbaObjectType     = new DeclaredType(IdlType::tk_objref,
							 0, 0);
    initialised_ = 1;
  }
}
